lib/active_record/connection_adapters/fibered_mysql2_adapter.rb

```diff
@@ -1,82 +1,72 @@
 # frozen_string_literal: true
 
-require 'em-synchrony'
 require 'active_model'
 require 'active_record/errors'
-require 'active_record/connection_adapters/em_mysql2_adapter'
+require 'active_record/connection_adapters/mysql2_adapter'
+require_relative '../../fibered_mysql2/async_task'
 
 module FiberedMysql2
-  module FiberedMysql2Adapter_4_2
+  module FiberedMysql2Adapter_6
     def lease
-      synchronize do
-        unless in_use?
-          @owner = Fiber.current
-        end
-      end
-    end
-  end
-
-  module FiberedMysql2Adapter_5_2
-    def lease
-      if in_use?
-        msg = "Cannot lease connection, ".dup
-        if owner_fiber == Fiber.current
-          msg << "it is already leased by the current fiber."
+      if (ot = owner_task)
+        msg = +"Cannot lease connection; "
+        if ot == (current_task = AsyncTask.current_or_none)
+          msg << "it is already leased by the current Async::Task."
         else
-          msg << "it is already in use by a different fiber: #{owner_fiber}. " \
-                  "Current fiber: #{Fiber.current}."
+          msg << "it is already in use by a different Async::Task: #{ot}. " \
+                  "Current Async::Task: #{current_task}."
         end
         raise ::ActiveRecord::ActiveRecordError, msg
       end
 
-      @owner = Fiber.current
+      @owner = AsyncTask.current_or_none
     end
 
     def expire
-      if in_use?
-        # Because we are actively releasing connections from dead fibers, we only want
-        # to enforce that we're expiring the current fibers connection, iff the owner
+      if (ot = owner_task)
+        # Because we are actively releasing connections from dead tasks, we only want
+        # to enforce that we're expiring the current task's connection, iff the owner
         # of the connection is still alive.
-        if owner_fiber.alive? && owner_fiber != Fiber.current
-          raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection, " \
-            "it is owned by a different fiber: #{owner_fiber}. " \
-            "Current fiber: #{Fiber.current}."
+        if ot.alive? && ot != (current_task = AsyncTask.current_or_none)
+          raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection; " \
+            "it is owned by a different Async::Task: #{ot}. " \
+            "Current Async::Task: #{current_task}."
         end
 
         @idle_since = ::Concurrent.monotonic_time
         @owner = nil
       else
-        raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection, it is not currently leased."
+        raise ::ActiveRecord::ActiveRecordError, "Cannot expire connection; it is not currently leased."
```

Comment:
Shouldn't `ActiveRecord::ConnectionNotEstablished` more appropriate error here?

Reply:
This code was adapted [straight from Rails](https://github.com/rails/rails/blob/99cd347a55fd9bddb458afd003e0f50c5b2ba29a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L254), so I wanted to keep the exception signatures the same.